File: src/main/java/algo/recursion/BinarySearch.java
package algo.recursion;

/**
 * Recursive binary search
 *
 * Worst time complexity - O(log n)
 * Worst space complexity - O(log n)
 *
 * todo: compare execution time with iterative search
 */
class BinarySearch {

  /**
   * Search for key in array of objects
   *
   * @param a   array of objects
   * @param key searched object
   * @return > 0 - position of the key in array
   * < 0 - if key is not found in array
   */
  static int search(Object[] a, Object key) {
    return searchRec(a, 0, a.length - 1, key);
  }

  private static int searchRec(Object[] a, int first, int last, Object key) {
    if (first <= last) {
      int mid = (first + last) >> 1;
      Comparable midVal = (Comparable) a[mid];
      int comp = midVal.compareTo(key);
      if (comp < 0)
        return searchRec(a, mid + 1, last, key);
      else if (comp > 0)
        return searchRec(a, first, mid - 1, key);
      else
        return mid;
    }
    return -first - 1;
  }
}

File: algo-python/algo/arrays/binary_search.py
# coding=utf-8
"""
Implementation of a binary search
"""


def binary_search(input_array, value):
    lo = 0
    hi = len(input_array) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if input_array[mid] < value:
            lo = mid + 1
        elif input_array[mid] > value:
            hi = mid - 1
        else:
            return mid
    return -1


test_list = [1, 3, 9, 11, 15, 19, 29]

print(binary_search(test_list, 25))
print(binary_search(test_list, 15))
print(binary_search(test_list, 29))
